addons/mail/static/src/utils/common/pdf_thumbnail.js

```diff
@@ -0,0 +1,55 @@
+import { loadPDFJSAssets } from "@web/core/utils/pdfjs";
+
+export async function generatePdfThumbnail(pdfUrl, options = { width: 180, height: 110, base64: true}) {
+    let initialWorkerSrc = false;
+    try {
+        await loadPDFJSAssets();
+        // Force usage of worker to avoid hanging the tab.
+        initialWorkerSrc = globalThis.pdfjsLib.GlobalWorkerOptions.workerSrc;
+        globalThis.pdfjsLib.GlobalWorkerOptions.workerSrc =
+            "/web/static/lib/pdfjs/build/pdf.worker.js";
+    } catch {
+        return { thumbnail: undefined, pdfEnabled: false };
+    }
+    // Force usage of worker to avoid hanging the tab.
+    initialWorkerSrc = globalThis.pdfjsLib.GlobalWorkerOptions.workerSrc;
+    globalThis.pdfjsLib.GlobalWorkerOptions.workerSrc = "/web/static/lib/pdfjs/build/pdf.worker.js";
+    let canvas;
+    try {
+        // Support for blob url
+        if (pdfUrl.startsWith("blob:")) {
+            pdfUrl = URL.createObjectURL(pdfUrl);
```

Comment:
I would revoke that object URL down the line (maybe in the finally?)

https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL_static